src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CppUnitTest.h"
+
+#include "ModernApp.h"
+#include "TestUtils.h"
+
+#include "UiaOperationAbstraction.h"
+
+using namespace UiaOperationAbstraction;
+
+namespace UiaOperationAbstractionTests
+{
+    // Detect whether a type can be returned from a remote operation. A user
+    // passes a single type parameter T in to CanBeReturned, and the type's
+    // underlying boolean value will represent whether the type can be returned
+    // or not.
+    //
+    // This works using a behavior called SFINAE. We define two overloads for
+    // CanBeReturned: one that inherits from true_type and one that inherits
+    // from false_type. When C++ is deciding which overload to choose, it
+    // starts with the true_type one, since it has the most specific signature.
+    // If the overload isn't ill-formed (i.e. the expression in decltype
+    // successfully compiles), the compiler will select that overload.
+    // Otherwise, it will move on to the false_type overload, which is never
+    // ill-formed. This means that the boolean value of the type is known at
+    // compile-time, and can be compared against an expected value.
+    //
+    // We use FromRemoteResult to test for whether the type can be returned
+    // because while calling BindResult on an un-returnable type would also
+    // fail to compile, the compiler can't detect compilation failures within
+    // that function. It can only reason about the flat expression in decltype.
+    template<class, class = std::void_t<>>
+    struct CanBeReturned : std::false_type{};
+
+    template<class T>
+    struct CanBeReturned<T,
+            std::void_t<decltype(std::declval<T>().FromRemoteResult(std::declval<winrt::Windows::Foundation::IInspectable>()))>
+        > : std::true_type{};
+
+    UiaOperationAbstraction::UiaCacheRequest CreateCacheRequest(const bool useRemoteOperations)
+    {
+        if (useRemoteOperations)
+        {
+            UiaOperationScope::GetCurrentDelegator()->NewRemoteCacheRequest();
+        }
```

Comment:
Hmm, maybe I'm missing something here, but this feels like it should be part of the abstraction library itself.

The same way how we don't need users to manually figure out how to initialize a `UiaArray`, we should have `UiaCacheRequest` initialize itself correctly too. Then we wouldn't need this `NewRemoteCacheRequest` method--again, same as for `UiaArray`.

Could this work?

Reply:
Yeah, I think you're right. Now that I'm thinking about it though it also raises the question of what are "supported" ways to create a cache request. In theory, we shouldn't allow it to be constructed from an IUIAutomationCacheRequest, since that cache request may contain fields (i.e. tree scope, etc.) which we don't support. It allows a weird backdoor to get around the restrictions on the UiaCacheRequest type.

I'll think about that some more.

Reply:
I'm using this thread to reply generally to the multiple comments by @mlalic, @mwcampbell, and @LukaszMakar. I was thinking about how to create a UiaCacheRequest overnight, and I believe the best approach is probably a default constructor.

Allowing a constructor that accepts IUIAutomationCacheRequest doesn't make sense, since that local type may have extra state (such as tree scope) which we don't support. It also just isn't really necessary to have an AutomationRemoteCacheRequest constructor since it doesn't need to be used.

We basically always need to create an empty cache request which determines its remoteness based on the current delegator. It needs to be initialized local or remote and never change. This is slightly different from UiaArray because UiaArray can be instantiated without regard for the remoteness of the current operation. UiaCacheRequest can't change remoteness, so it needs to be initialized correctly.

tl;dr I think the solution is to make the only way to construct a UiaCacheRequest a default initializer that reads the current delegator's remoteness value.

Reply:
This reasoning sounds sensible to me. I also like doing the simpler thing (just the one ctor) until there's a clear need for more, especially when from everything that we see right now, we can't find a good reason anyway.

---
In reply to: [459790094](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459790094) [](ancestors = 459790094)

Reply:
Agreed. I expressed my concerns in another comment. This feels fine to me. :)

---
In reply to: [459798282](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r459798282) [](ancestors = 459798282,459790094)